xmppserver/src/main/java/org/jivesoftware/openfire/websocket/XmppWebSocket.java

```diff
@@ -96,7 +97,8 @@ public void onConnect(Session session)
     @OnWebSocketClose
     public void onClose(int statusCode, String reason)
     {
-        closeSession();
+        // Handle asynchronously, to prevent deadlocks. See OF-2473.
+        HttpBindManager.getInstance().getSessionManager().execute(this::closeSession);
```

Comment:
You'll need to wrap this in a try/catch to ensure any errors during processing are properly logged and not lost in the ether. Either in this call, or in closeSession itself.

Reply:
Oh, absolutely! Losing errors like that can be very confusing. I've modified the asynchronous call. Thanks!